src/xrpld/app/tx/detail/LoanPay.cpp

```diff
@@ -0,0 +1,583 @@
+    }
+
+    if (auto const ret = checkFrozen(ctx.view, account, asset))
+    {
+        JLOG(ctx.j.warn()) << "Borrower account is frozen.";
+        return ret;
+    }
+    if (auto const ret = checkDeepFrozen(ctx.view, vaultPseudoAccount, asset))
+    {
+        JLOG(ctx.j.warn())
+            << "Vault pseudo-account can not receive funds (deep frozen).";
+        return ret;
+    }
+    if (auto const ret = requireAuth(ctx.view, asset, account))
+    {
+        JLOG(ctx.j.warn()) << "Borrower account is not authorized.";
+        return ret;
+    }
+    // Make sure the borrower has enough funds to make the payment!
+    // Do not support "partial payments" - if the transaction says to pay X,
+    // then the account must have X available, even if the loan payment takes
+    // less.
+    if (auto const balance = accountCanSend(
+            ctx.view,
+            account,
+            asset,
+            fhZERO_IF_FROZEN,
+            ahZERO_IF_UNAUTHORIZED,
+            ctx.j);
+        balance < amount)
+    {
+        JLOG(ctx.j.warn()) << "Payment amount too large. Amount: "
+                           << to_string(amount.getJson())
+                           << ". Balance: " << to_string(balance.getJson());
+        return tecINSUFFICIENT_FUNDS;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+LoanPay::doApply()
+{
+    auto const& tx = ctx_.tx;
+    auto& view = ctx_.view();
+
+    auto const amount = tx[sfAmount];
+
+    auto const loanID = tx[sfLoanID];
+    auto const loanSle = view.peek(keylet::loan(loanID));
+    if (!loanSle)
+        return tefBAD_LEDGER;  // LCOV_EXCL_LINE
+    std::int32_t const loanScale = loanSle->at(sfLoanScale);
+
+    auto const brokerID = loanSle->at(sfLoanBrokerID);
+    auto const brokerSle = view.peek(keylet::loanbroker(brokerID));
+    if (!brokerSle)
+        return tefBAD_LEDGER;  // LCOV_EXCL_LINE
+    auto const brokerOwner = brokerSle->at(sfOwner);
+    auto const brokerPseudoAccount = brokerSle->at(sfAccount);
+    auto const vaultID = brokerSle->at(sfVaultID);
+    auto const vaultSle = view.peek(keylet::vault(vaultID));
+    if (!vaultSle)
+        return tefBAD_LEDGER;  // LCOV_EXCL_LINE
+    auto const vaultPseudoAccount = vaultSle->at(sfAccount);
+    auto const asset = *vaultSle->at(sfAsset);
+
+    // Determine where to send the broker's fee
+    auto coverAvailableProxy = brokerSle->at(sfCoverAvailable);
+    TenthBips32 const coverRateMinimum{brokerSle->at(sfCoverRateMinimum)};
+    auto debtTotalProxy = brokerSle->at(sfDebtTotal);
+
+    // Send the broker fee to the owner if they have sufficient cover available,
+    // _and_ if the owner can receive funds. If not, so as not to block the
+    // payment, add it to the cover balance (send it to the broker pseudo
+    // account).
+    //
+    // Normally freeze status is checked in preflight, but we do it here to
+    // avoid duplicating the check. It'll claim a fee either way.
+    bool const sendBrokerFeeToOwner = [&]() {
```

Comment:
The lambda is not needed. Can just pass `Number::upward` to `roundToAsset()`.

Reply:
> The lambda is not needed. Can just pass `Number::upward` to `roundToAsset()`.

No, it can't, because only `roundToAsset` will be altered that way. We also need to capture the rounding up for `tenthBipsOfValue`. Using the lambda scope gets both of them in one shot.